mage_ai/streaming/sources/nats_js.py

```diff
@@ -0,0 +1,175 @@
+import asyncio
+import json
+import ssl
+import threading
+from dataclasses import dataclass, field
+from typing import Callable, Dict, List, Optional
+
+import nats
+from nats.errors import NoServersError
+
+from mage_ai.shared.config import BaseConfig
+from mage_ai.streaming.constants import DEFAULT_BATCH_SIZE, DEFAULT_TIMEOUT_MS
+from mage_ai.streaming.sources.base import BaseSource
+
+
+@dataclass
+class SSLConfig:
+    cafile: str = None
+    certfile: str = None
+    keyfile: str = None
+    check_hostname: bool = False
+
+
+@dataclass
+class NATSConfig(BaseConfig):
+    server_url: str
+    stream_name: str
+    subject: str = None
+    subjects: List = field(default_factory=list)
+    user_credentials: Optional[str] = None
+    use_tls: bool = False
+    ssl_config: SSLConfig = None
+    consumer_name: str = None
+    batch_size: int = DEFAULT_BATCH_SIZE
+    timeout: int = DEFAULT_TIMEOUT_MS / 1000  # Convert to seconds
+
+    @classmethod
+    def parse_config(self, config: Dict) -> Dict:
+        ssl_config = config.get('ssl_config')
+        if ssl_config and type(ssl_config) is dict:
+            config['ssl_config'] = SSLConfig(**ssl_config)
+        return config
+
+
+class NATSSource(BaseSource):
+    config_class = NATSConfig
+
+    def __init__(self, config, **kwargs):
+        self.loop = asyncio.new_event_loop()
+        self.thread = threading.Thread(target=self.start_loop, daemon=True)
+        self.thread.start()
+        super().__init__(config)
+
+    def start_loop(self):
+        asyncio.set_event_loop(self.loop)
+        self.loop.run_forever()
+
+    def stop_loop(self):
+        self.loop.call_soon_threadsafe(self.loop.stop)
+        self.thread.join()
+
+    async def ainit_client(self):
+        try:
+            connect_opts = {
+                "servers": [self.config.server_url],
+                "error_cb": self.error_cb,
+                "reconnected_cb": self.reconnected_cb,
+                "disconnected_cb": self.disconnected_cb,
+                "closed_cb": self.closed_cb,
+            }
+
+            # Configure SSL context if use_tls is True
+            if self.config.use_tls and self.config.ssl_config:
+                ssl_ctx = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
+                if self.config.ssl_config.cafile:
+                    ssl_ctx.load_verify_locations(self.config.ssl_config.cafile)
+                if self.config.ssl_config.certfile and self.config.ssl_config.keyfile:
+                    ssl_ctx.load_cert_chain(
+                        certfile=self.config.ssl_config.certfile,
+                        keyfile=self.config.ssl_config.keyfile
+                    )
+                connect_opts["tls"] = ssl_ctx
+
+            # Use NKEY if provided
```

Comment:
The `user_credentials` param is relevant when using decentralized auth. Checkout the available auth options here: https://github.com/nats-io/nats.py/blob/main/nats/aio/client.py#L310-L318.

Reply:
Was thinking of just supporting NKEYs for this iteration, I should just have to pass both of the NKEYs correct?

Reply:
In this case, it would be `nkeys_seed` (as a file) or `nkeys_seed_str` for passing a string directly.

Reply:
addressed with 56517c94e